src/session/client-certificate.c

```diff
@@ -49,31 +93,55 @@
  * including "0::" prefix and newline.
  * NB: the kernel doesn't allow newlines in cgroup names. */
 static char *
-read_proc_self_cgroup (size_t *out_length)
+read_proc_pid_cgroup (int dirfd, size_t *out_length)
 {
-  FILE *fp = fopen ("/proc/self/cgroup", "r");
-
-  if (fp == NULL)
-    {
-      warn ("Failed to open /proc/self/cgroup");
-      return NULL;
-    }
-
   char buffer[1024];
-  *out_length = fread (buffer, 1, sizeof (buffer) - 1, fp);
-  buffer[*out_length] = '\0';
-  fclose (fp);
+  ssize_t len = read_proc_file (dirfd, "cgroup", buffer, sizeof buffer);
+  if (len < 0)
+      return NULL;
 
-  if (*out_length > 5 &&  /* at least "0::/\n" */
-      *out_length <= sizeof (buffer) - 1 && /* not too big */
+  if (len > 5 &&  /* at least "0::/\n" */
+      len <= sizeof (buffer) - 1 && /* not too big */
       strncmp (buffer, "0::", 3) == 0 && /* must be a cgroupsv2 */
```

Comment:
I feel like this is the only one which is actually necessary: it will also implicitly assert that the string is long enough, won't it?  Maybe add the `/` here, if that's part of what you're definitely expecting.

So like

```c
if (strncmp(buffer, "0::/", 4) == 0 && buffer[len - 1] == '\n')
```

Although I kinda get if you want to be more explicit about checking that len is at least 5.  That index looks a bit unguarded otherwise.

But: why didn't we just add this check to the `read_proc_file()` code?  It would be nice to share it for the stat case too, and since we're trying to detect the case where we did a short read, we can be certain that we'll see the newline at the end in both cases.

Reply:
Right, I simplified the check. However, I don't want to move the `\n` check to `read_proc_file()` -- I don't want to (nor have to) assume that all files in /proc are \n terminated. We specifically depend on the `\n` for the cgroup, but not for the  `/stat` parsing.

Reply:
Are you trying to guard against a future where `/proc/*/stat` is not newline-terminated?  I know you had an argument before about "maybe the file got a lot larger" but we're already not capable of dealing that because of our check for incomplete-file-reading.